src/core/mac/sub_mac.cpp

```diff
@@ -990,6 +991,27 @@ void SubMac::StartTimerAt(Time aStartTime, uint32_t aDelayUs)
 
 // LCOV_EXCL_START
 
+struct SubMac::StateValueChecker
+{
+    StaticCounterInit(0);
+
+    CheckEnum(kStateDisabled, "kStateDisabled value is not correct");
+    CheckEnum(kStateSleep, "kStateSleep value is not correct");
+    CheckEnum(kStateReceive, "kStateReceive value is not correct");
+    CheckEnum(kStateCsmaBackoff, "kStateCsmaBackoff value is not correct");
+    CheckEnum(kStateTransmit, "kStateTransmit value is not correct");
+    CheckEnum(kStateEnergyScan, "kStateEnergyScan value is not correct");
+#if OPENTHREAD_CONFIG_MAC_ADD_DELAY_ON_NO_ACK_ERROR_BEFORE_RETRY
+    CheckEnum(kStateDelayBeforeRetx, "kStateDelayBeforeRetx value is not correct");
+#endif
+#if !OPENTHREAD_MTD && OPENTHREAD_CONFIG_MAC_CSL_TRANSMITTER_ENABLE
+    CheckEnum(kStateCslTransmit, "kStateCslTransmit value is not correct");
+#endif
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
+    CheckEnum(kStateCslSample, "kStateCslSample value is not correct");
+#endif
+};
```

Comment:
One thing which may not allow this is that we are defining `const int value_{LINE}` which we may not be able to do in a local (nested within a function/method) type definition. But if we can do https://github.com/openthread/openthread/pull/10843#discussion_r1803633411 then this may work?